### PR TITLE
Support phe algorithm EC-Elgamal

### PR DESCRIPTION
**Pull request by jinjiu**

### CODE REVIEW

**Review by InfoHunter — COMMENTED** [review]
缺EVP的test，好像不能从EVP的路径调用ElGamal生成密钥或者加解密？

**Comment by InfoHunter** [comment]
doc目录下，缺ElGamal的文档

File: util/libcrypto.num [hunk 1]
```diff
@@ -4648,3 +4648,14 @@ EVP_KDF_vctrl                           6594	1_1_1b	EXIST::FUNCTION:
 EVP_KDF_ctrl_str                        6595	1_1_1b	EXIST::FUNCTION:
 EVP_KDF_size                            6596	1_1_1b	EXIST::FUNCTION:
 EVP_KDF_derive                          6597	1_1_1b	EXIST::FUNCTION:
+EC_ELGAMAL_encrypt                      6598	1_1_1h	EXIST::FUNCTION:EC,EC_ELGAMAL
```
Comment (InfoHunter) [inline util/libcrypto.num:4651]: 我理解这种应该不需要特殊处理，除非EVP框架包含不了ElGamal的某些特殊参数，否则是可以纳入到EVP里的

**Review by KaneX — COMMENTED** [review]
Comment added in ec_elgamal.c, concerning the timing side channel of discrete logarithm.

File: crypto/ec/ec_elgamal.c [hunk 2]
```diff
@@ -0,0 +1,748 @@
...
+    OPENSSL_free(entry->key);
+    OPENSSL_free(entry);
+}
+
+/** Creates a new EC_ELGAMAL_BSGS_HASH_TABLE object
+ *  \param  ctx   EC_ELGAMAL_CTX object
+ *  \param  size  the size of the ecdlp bsgs hash table
+ *  \return newly created EC_ELGAMAL_BSGS_HASH_TABLE object or NULL in case of an error
+ */
+static EC_ELGAMAL_BSGS_HASH_TABLE *EC_ELGAMAL_BSGS_HASH_TABLE_new(EC_ELGAMAL_CTX *ctx,
```
[the remark this answers is not in the document]
Comment (InfoHunter) [reply crypto/ec/ec_elgamal.c:204]: How possible is it for an adversary to exploit this?

[the remark this answers is not in the document]
Comment (KaneX) [reply crypto/ec/ec_elgamal.c:204]: Depends on the application. 

Say Alice encrypts m1 and sends Enc(m1) to Eve, Eve computes Enc(m1+m1) and sends to Alice for decryption, then Eve asks Alice to send Enc(m2) for the next homomorphic computation. If Alice sends Enc(m2) immediately then Eve can guess the size of m1 by recording the time, and Eve is playing honest in this attack. Imagine that m1 is someone's salary, then it is possible that Eve gets something.

If Eve cannot record the time of decryption, say there is no Enc(m2) sent in this scenario, then it is hard to carry out this attack. If Eve can repeatedly record the time (for example in some for-loop), then she may accurately guess someone's salary.

File: util/libcrypto.num (hunk 1 above)
Comment (jinjiu) [reply util/libcrypto.num:4651]: 你的意思是这些接口不需要暴露出来，直接纳入到 EVP 框架里是吗？

Comment (InfoHunter) [reply util/libcrypto.num:4651]: 目前看不需要了，因为要设计新API

**Review by InfoHunter — APPROVED** [review]

File: crypto/ec/ec_elgamal.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (InfoHunter) [reply crypto/ec/ec_elgamal.c]: It seems we could merge this PR without resolving this potential problem at current stage. Meanwhile, we can open a new issue to track this for further improvement.

**Review by KaneX — APPROVED** [review]

File: crypto/ec/ec_elgamal.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (KaneX) [reply crypto/ec/ec_elgamal.c]: I agree.

**Pull request merged**